### PR TITLE
Add support for allowing a site subset to be native AMP

### LINKED ISSUE #934: Add theme support scenarios: allow paired mode to serve some content exclusively in AMP
The plugin in `amp` theme support currently supports two modes:

1. Canonical: Everything is canonical AMP.
2. Paired: Everything is available in non-AMP and _some_ content is also available in AMP, with _different_ templates used in AMP mode.

There is a third case that should be supported and that is:

3. Mixed-Some-Optionally-AMP: All content is available in non-AMP, and some content is also available in AMP, but they use the _same theme templates_. 

Now that we have full-document sanitization in #929, the a non-AMP template can actually be served as totally valid AMP, with `amp` attribute added to the `html` element, meta viewport being added, and CSS boilerplate code. So we can actually use the exact same templates in AMP as non-AMP. We don't readily make this easy in paired mode, as to do paired mode the `amp` theme support must define a `template_dir` for where to locate the AMP templates, like:

```php
add_theme_support( 'amp', array(
    'template_dir' => 'amp-templates/'
) );
```

When a paired-mode site only allows a subset of content to be shown in AMP (e.g. just singular posts), they can explicitly indicate this with the `available_callback` property:

```php
add_theme_support( 'amp', array(
    'template_dir' => 'amp-templates/'
    'available_callback' => 'is_single',
) );
```

However, in order to support re-using templates between AMP and non-AMP we should also be able to do:

```php
add_theme_support( 'amp', array(
    'available_callback' => 'is_single',
) );
```

Without this, you have to do a needless hack of setting the `template_dir` to `.` or add a `single.php` in the template directory that just does `require __DIR__ . '../single.php';`

Lastly, there is a fourth mode that should be doable:

4. Mixed-Some-Only-AMP: Some content is not available in AMP, and everything else is _only_ available in AMP.

In the fourth mode there should not be two versions of any one URL. There would be no `/amp/` endpoints, and no `amphtml` links on AMP content. Content in AMP and non-AMP would be disjoint sets. To do this, there would need to probably need to be something different from `available_callback` in the theme support, or rather instead of this callback returning a boolean flag it should actually be tri-state:

* AMP Omitted
* AMP Optional
* AMP Only

For example, if you wanted to serve AMP-only posts, AMP optionally for pages, and omit AMP everywhere else, you could do:

```php
add_theme_support( 'amp', array(
    'available_callback' => function() {
        if ( is_single() ) {
            return AMP_ONLY; // No amphtml version offered. AMP 100%.
        } elseif (  is_page()) {
            return AMP_OPTIONAL; // Non-AMP by default. Link to amphtml URL offered to opt-in.
        } else {
            return AMP_OMITTED; // Non-AMP always. No link to amphtml URL offered at all.
        }
    }
) );
```

### PR DESCRIPTION
**Pull request by westonruter**
It's been a journey trying to figure out the best way to empower both end users and theme developers over which parts of their site are served as AMP, and whether AMP is served as canonical or is available in paired mode.

ℹ️  The commits in this PR will be squashed/rebased when merged, so please excuse the messy commit history in this branch.

When in native/paired template modes, the Post Types section of the admin screen is extended to also display a selection of the templates for which AMP can be served on the site:

image

The supportable templates are organized hierarchically according to the WordPress template hierarchy, so you enable the Category template for AMP but the general Archives template is disabled, then the Category template status will override the Archives template status.

The available templates can be amended via the `amp_supportable_templates` filter. This filter can also be used by a theme/plugin to force AMP to be enabled or disabled for a given template, such as here to forbid AMP from the Category templates (since maybe they have some required custom JS):

```php
add_filter( 'amp_supportable_templates', function( $templates ) {
	$templates['is_category']['supported'] = false;
	return $templates;
} );
```

Or to force AMP to be enabled for the author template:

```php
add_filter( 'amp_supportable_templates', function( $templates ) {
	$templates['is_author']['supported'] = false;
	return $templates;
} );
``` 

You can use this filter to add new templates as well. For example if you wanted to give the user control over the year archive template (take note of the `parent`):

```php
add_filter( 'amp_supportable_templates', function( $templates ) {
	$templates['is_year'] = array(
		'label'  => __( 'Year', 'example' ),
		'parent' => 'is_date',
	);
	return $templates;
} );
```

That then is presented as:

image

Lastly, there is a checkbox to automatically serve all templates and post types as AMP:

image

The availability of this option may will be determined whether or not a theme/plugin is forcibly setting the supported status for a given template.

You may also force all templates to be served as AMP at the theme support level:

```php
add_theme_support( 'amp', array(
    'templates_supported' => 'all'
) );
```

Or force certain templates to be supported or unsupported:

```php
add_theme_support( 'amp', array(
    'templates_supported' => array(
        'is_category' => false,
        'is_author' => true,
    )
) );
```

# Changelog Summary

* The template hierarchy is now exposed in the admin screen allowing the user to specify which kinds of templates should be served as AMP.
* The `amp_supportable_templates` filter allows you to customize the template hierarchy, add recognition for new template types, and force certain templates to be supported or forbidden in AMP.
* There is a new `templates_supported` arg on the `amp` theme support which can take a value of 'all' to force AMP to be available on all templates, or else an array mapping template IDs (e.g. `is_search`) to whether they are supported.
* The “Post” post type can now be turned off instead of always being forcibly enabled. 
* When `amp` theme support is present, the static homepage and page for posts are both enabled by default, as are pages with custom templates assigned.
* The template mode is now explicitly indicated via a `mode` theme support arg, which can either be `native` or `paired`.
* The `template_dir` can now be used in native mode as well.
* You can now define `add_theme_support( 'amp', array( 'mode' => 'native', 'optional' => true ) )` where the `optional` flag allows you to bypass actually activating the support in code, but forcing the mode to be `native` if they do add support. This allows the user to define `comments_live_list` in the theme without forcing AMP support to be present.
* Loading AMP component scripts has been restored on a native-mode site where AMP content appears on a non-AMP page.
* The `post_supports_amp()` logic has been consolidated into `AMP_Post_Type_Support::get_support_errors()`.
* ❗️ Breaking change: The `available_callback` has been eliminated.
* The restriction that `is_amp_endpoint()` cannot be called prior to `parse_query` action has been restored.

# Todo

- [x] Determine how a theme can force AMP to be used on all templates. If `mode` is `native` and AMP components are being used in all templates, then the theme needs to force all responses to be AMP. This could be done by a theme setting `supported` to true for all the `amp_supportable_templates`, but it seems there should be an easier way.
- [x] Add an “other” option to determine if AMP gets served by default for any templates not explicitly called out. This would be used in particular for `template_redirect` templates that are not loaded normally.
- [x] Address todos.
- [x] Should there be a way for a theme to explicitly mark templates as requiring for AMP other than by using the filter? An `all_templates_supported` flag on the theme support as a shortcut for using the filter? There could be a `templates_supported` arg when defining theme support which can take either `'all'` or an array of the template conditions which then become immutable.
- [x] Restore facilitation of serving AMP components in non-AMP documents if mode is native.

Fixes #934.

### CODE REVIEW

File: includes/amp-helper-functions.php [hunk 1]
```diff
@@ -239,80 +239,62 @@ function amp_add_amphtml_link() {
...
 	}
 
-	return false;
+	if ( current_theme_supports( 'amp' ) ) {
+		$template_available = true === AMP_Theme_Support::get_template_availability();
+		return amp_is_canonical() ? $template_available : $has_amp_query_var;
+	} else {
+
+		// Check if the queried object supports AMP.
+		if ( is_singular() || ( $wp_query instanceof WP_Query && $wp_query->is_posts_page ) ) {
```
Comment (kienstra) [inline includes/amp-helper-functions.php]: Good idea to check that `$wp_query->is_posts_page`

File: includes/class-amp-theme-support.php [hunk 2]
```diff
@@ -133,26 +141,56 @@ public static function init() {
 		add_action( 'wp', array( __CLASS__, 'finish_init' ), PHP_INT_MAX );
 	}
 
+	/**
+	 * Determine whether theme support as added via option.
```
Comment (kienstra) [inline includes/class-amp-theme-support.php]: A very minor point, but this might be:
>theme support **was** added

File: includes/class-amp-theme-support.php [hunk 3]
```diff
@@ -303,13 +337,246 @@ public static function is_customize_preview_iframe() {
...
+		 *
+		 * A theme or plugin can force a given template to be supported or not by preemptively
+		 * setting the 'supported' flag for a given template. Otherwise, if the flag is undefined
+		 * then the user will be able to toggle it themselves in the admin.
+		 *
+		 * @since 1.0
+		 *
+		 * @param array $templates Supportable templates.
+		 */
+		$templates = apply_filters( 'amp_supportable_templates', $templates );
```
Comment (kienstra) [inline includes/class-amp-theme-support.php]: This filter is already well-documented.

 But maybe it could also mention that each `$template` in `$templates` [can have](https://github.com/Automattic/amp-wp/pull/1235/files#diff-2585472f207548f626a43a7ff3cab922R386) an `$id` or `'callback'` that is a method of `WP_Query` (if I understand this right). Or it can simply be a `callable`.

File: includes/options/class-amp-options-menu.php [hunk 4]
```diff
@@ -272,45 +275,148 @@ public function render_validation_handling() {
...
+			<ul>
+			<?php foreach ( array_map( 'get_post_type_object', AMP_Post_Type_Support::get_eligible_post_types() ) as $post_type ) : ?>
+				<li>
+					<?php $element_id = AMP_Options_Manager::OPTION_NAME . "-supported_post_types-{$post_type->name}"; ?>
+					<input
+						type="checkbox"
+						id="<?php echo esc_attr( $element_id ); ?>"
+						name="<?php echo esc_attr( $element_name ); ?>"
+						value="<?php echo esc_attr( $post_type->name ); ?>"
+						<?php checked( true, post_type_supports( $post_type->name, amp_get_slug() ) ); ?>
```
Comment (kienstra) [inline includes/options/class-amp-options-menu.php]: The first argument of `checked()` probably isn't needed.

File: includes/options/class-amp-options-menu.php [hunk 5]
```diff
@@ -272,45 +275,148 @@ public function render_validation_handling() {
...
+			<p class="description">
+				<?php esc_html_e( 'This will allow all of the URLs on your site to be served as AMP by default.', 'amp' ); ?>
+			</p>
+		</fieldset>
+
+		<fieldset id="singular_templates_fieldset">
+			<?php $element_name = AMP_Options_Manager::OPTION_NAME . '[supported_post_types][]'; ?>
+			<h4 class="title"><?php esc_html_e( 'Content Types', 'amp' ); ?></h4>
+			<p>
+				<?php esc_html_e( 'The following content types will be available as AMP by default, but you can override this on an item-by-item basis:', 'amp' ); ?>
```
Comment (kienstra) [inline includes/options/class-amp-options-menu.php]: Maybe this would be enough:
>The following content types will be available as AMP:

If they're checked by default, that might communicate "will be available as AMP **by default**"

content-types

File: includes/class-amp-theme-support.php [hunk 6]
```diff
@@ -303,13 +337,246 @@ public static function is_customize_preview_iframe() {
...
+	/**
+	 * Get the templates which can be supported.
+	 *
+	 * @return array Supportable templates.
+	 */
+	public static function get_supportable_templates() {
+		$templates = array(
+			'is_singular' => array(
+				'label'       => __( 'Singular', 'amp' ),
+				'description' => __( 'Required for the above content types.', 'amp' ),
```
Comment (kienstra) [inline includes/class-amp-theme-support.php]: clicking-content-type

File: includes/class-amp-theme-support.php [hunk 7]
```diff
@@ -118,7 +126,7 @@ public static function init() {
 			$args = array_shift( $support );
 			if ( ! is_array( $args ) ) {
 				trigger_error( esc_html__( 'Expected AMP theme support arg to be array.', 'amp' ) ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_trigger_error
```
Comment (kienstra) [inline includes/class-amp-theme-support.php]: This appears on the AMP Settings page `/wp-admin/admin.php?page=amp-options`
>( ! ) Warning: array_merge(): Argument #2 is not an array in /srv/www/loc/public_html/wp-content/plugins/amp/includes/class-amp-theme-support.php on line 189
>( ! ) Notice: Expected AMP theme support arg to be array. in /srv/www/loc/public_html/wp-content/plugins/amp/includes/class-amp-theme-support.php on line 128

This occurs with Twenty Seventeen. That would make sense, as it doesn't do `add theme_support( 'amp' )`, so `$support` is `array()`

And `$args` is null, because `$args = array_shift( $support );`

File: includes/options/class-amp-options-menu.php (hunk 5 above)
Comment (westonruter) [reply includes/options/class-amp-options-menu.php]: Actually they would be available as AMP by default, no?

**Comment by rheinardkorf** [comment]
@westonruter works really well with custom templates in plugins as well.  Thanks for doing this.

**Comment by westonruter** [comment]
Ready for review and testing. With this merged, I believe we're ready for beta.

File: amp.php [hunk 8]
```diff
@@ -272,28 +272,53 @@ function amp_correct_query_when_is_front_page( WP_Query $query ) {
...
 function amp_is_canonical() {
-	$support = get_theme_support( 'amp' );
-	if ( true === $support ) {
-		return true;
+	if ( ! current_theme_supports( 'amp' ) ) {
+		return false;
 	}
+
+	$mode    = 'native';
+	$support = get_theme_support( 'amp' );
```
Comment (amedina) [inline amp.php]: May there be a more differentiating naming for `current_theme_supports` vs. `get_theme_support`? Probably I am missing some of the context but the naming appear a little confusing.

File: includes/admin/class-amp-editor-blocks.php [hunk 9]
```diff
@@ -43,6 +43,22 @@ public function init() {
 		if ( function_exists( 'gutenberg_init' ) ) {
 			add_action( 'enqueue_block_editor_assets', array( $this, 'enqueue_block_editor_assets' ) );
 			add_filter( 'wp_kses_allowed_html', array( $this, 'whitelist_block_atts_in_wp_kses_allowed_html' ), 10, 2 );
+
+			/*
+			 * Dirty AMP is required when a site is in native mode but not all templates are being served
```
Comment (amedina) [inline includes/admin/class-amp-editor-blocks.php]: This comment is hard to follow; perhaps not mentioning "Dirty AMP" but explaining directly the semantics of the filter and action functions may be clearer. Just a thought.

File: amp.php (hunk 8 above)
Comment (westonruter) [reply amp.php]: The `current_theme_supports()` and `get_theme_support()` are WordPress core functions. They are somewhat confusing. The former returns whether support has been added (boolean), _and_ applies a filter to let plugins control whether support is present. The `get_theme_support()` on the other hand does not apply a filter, but rather it returns the args that were passed in for the original `add_theme_support()` call.

File: includes/admin/class-amp-post-meta-box.php [hunk 10]
```diff
@@ -143,12 +143,20 @@ public function enqueue_admin_assets() {
...
+			$support_errors = AMP_Post_Type_Support::get_support_errors( $post );
+		}
+
 		wp_add_inline_script( self::ASSETS_HANDLE, sprintf( 'ampPostMetaBox.boot( %s );',
 			wp_json_encode( array(
 				'previewLink'     => esc_url_raw( add_query_arg( amp_get_slug(), '', get_preview_post_link( $post ) ) ),
 				'canonical'       => amp_is_canonical(),
-				'enabled'         => post_supports_amp( $post ),
-				'canSupport'      => count( AMP_Post_Type_Support::get_support_errors( $post ) ) === 0,
+				'enabled'         => empty( $support_errors ),
```
Comment (amedina) [inline includes/admin/class-amp-post-meta-box.php]: Very nice semantics.

File: includes/admin/class-amp-post-meta-box.php [hunk 11]
```diff
@@ -170,23 +178,32 @@ public function render_status( $post ) {
...
-			! amp_is_canonical()
 		);
 
 		if ( true !== $verify ) {
 			return;
 		}
 
-		$errors = AMP_Post_Type_Support::get_support_errors( $post );
-		$status = post_supports_amp( $post ) ? self::ENABLED_STATUS : self::DISABLED_STATUS;
+		if ( current_theme_supports( 'amp' ) ) {
```
Comment (amedina) [inline includes/admin/class-amp-post-meta-box.php]: Would it be possible having a comment for this `if-then-else`?

File: includes/options/class-amp-options-menu.php (hunk 5 above)
Comment (kienstra) [reply includes/options/class-amp-options-menu.php]: Yes, they would be available by default. But I was thinking that it might not be necessary to mention that. 

But on second thought, that description is probably helpful. Most people probably don't want to disable AMP on posts, for example.

File: includes/admin/class-amp-post-meta-box.php (hunk 11 above)
Comment (westonruter) [reply includes/admin/class-amp-post-meta-box.php]: To explain why the condition is here?

File: includes/class-amp-theme-support.php (hunk 3 above)
Comment (kienstra) [reply includes/class-amp-theme-support.php]: Thanks, this looks good.

File: includes/options/class-amp-options-menu.php (hunk 4 above)
Comment (kienstra) [reply includes/options/class-amp-options-menu.php]: Thanks, this looks good.

**Comment by westonruter** [comment]
I realized that we don't need to remove `available_callback`. We can still support for legacy installs (0.7) it but mark it as deprecated. In short, we just need to add this to the top of `\AMP_Theme_Support::get_template_availability()`:

```php
		// Support available_callback from 0.7, though it is deprecated.
		if ( isset( $theme_support_args['available_callback'] ) && is_callable( $theme_support_args['available_callback'] ) ) {
			_doing_it_wrong( 'add_theme_support', esc_html__( 'The available_callback is deprecated when adding amp theme support in favor of declaratively setting the supported_templates.', 'amp' ), '1.0' );
			return call_user_func( $theme_support_args['available_callback'] );
		}
```

I'll do that.

File: includes/options/class-amp-options-menu.php [hunk 12]
```diff
@@ -272,45 +281,168 @@ public function render_validation_handling() {
...
+		<fieldset id="supported_post_types_fieldset">
+			<?php $element_name = AMP_Options_Manager::OPTION_NAME . '[supported_post_types][]'; ?>
+			<h4 class="title"><?php esc_html_e( 'Content Types', 'amp' ); ?></h4>
+			<p>
+				<?php esc_html_e( 'The following content types will be available as AMP:', 'amp' ); ?>
+			</p>
+			<ul>
+			<?php foreach ( array_map( 'get_post_type_object', AMP_Post_Type_Support::get_eligible_post_types() ) as $post_type ) : ?>
+				<li>
+					<?php $element_id = AMP_Options_Manager::OPTION_NAME . "-supported_post_types-{$post_type->name}"; ?>
```
Comment (kienstra) [inline includes/options/class-amp-options-menu.php]: **Media Pages In "Classic" Mode**

This isn't the most common use-case. But it looks like on the "Media" pages in "Classic" mode, the `href` in the `<link rel="amphtml">` leads to a 404 page. It ends with `/amp`, which results in a 404 page. But the `?amp` URL works.

Steps to reproduce:
1. Activate Twenty Sixteen
2. On the "AMP Settings" page, select "Template Mode" > "Classic," and check "Supported Templates" > "Media"
classic-media-templates
3. In the Media Library, click an image, and click "View Attachment Page":
media-page

4. In the source of that page, the `<link rel="amphtml">` has an `href` with `/amp`, not `?amp`:

```html
<link rel="amphtml" href="https://loc.test/lorem-2000bridge/amp/">
```
source-media-page

5. Expected: that URL is a valid AMP page
6. Actual: it's a 404 page:
ends-amp

7. Going to the same URL, but ending in `?amp` results in a valid AMP page:
ends-amp-query-var

Comment (kienstra) [inline includes/options/class-amp-options-menu.php]: **Media Pages In "Classic" Mode**

This isn't the most common use-case. But it looks like on the "Media" pages in "Classic" mode, the `href` in the `<link rel="amphtml">` leads to a 404 page. It ends with `/amp`, which results in a 404 page. But the `?amp` URL works.

Steps to reproduce:
1. Activate Twenty Sixteen
2. On the "AMP Settings" page, select "Template Mode" > "Classic," and check "Supported Templates" > "Media"
classic-media-templates
3. In the Media Library, click an image, and click "View Attachment Page":
media-page

4. In the source of that page, the `<link rel="amphtml">` has an `href` with `/amp`, not `?amp`:

```html
<link rel="amphtml" href="https://loc.test/lorem-2000bridge/amp/">
```
source-media-page

5. Expected: that URL is a valid AMP page
6. Actual: it's a 404 page:
ends-amp

7. Going to the same URL, but ending in `?amp` results in a valid AMP page:
ends-amp-query-var

This seems to exist even after flushing permalinks.

**Comment by westonruter** [comment]
Using `available_callback` now works again but it is deprecated, and the admin shows:

image

File: includes/options/class-amp-options-menu.php (hunk 12 above)
Comment (westonruter) [reply includes/options/class-amp-options-menu.php]: Good catch. I believe this is fixed after 7091b35.

Comment (kienstra) [reply includes/options/class-amp-options-menu.php]: Thanks, your commit [7091b35](https://github.com/Automattic/amp-wp/commit/7091b35ec7f316680cf5d1c6cf77a66dd6aa47b8) fixed this, and the `<link rel="amphtml">` now leads to a valid AMP page.

File: includes/class-amp-theme-support.php (hunk 6 above)
Comment (jwold) [reply includes/class-amp-theme-support.php]: @kienstra I like the idea of making it simpler for people, so that if they select a content type the singular template would be auto selected.

Is there ever a situation where I would select a singular template and NOT select a content type?

Comment (westonruter) [reply includes/class-amp-theme-support.php]: @jwold Yes, it would probably be common to have the Singular template checked but not have all Content Types checked. What would probably not be as common is to have Singular unchecked but the Content Types checked. There is a scenario where it could be used though: maybe the post archive page for a given CPT can be served as AMP but the singular CPT template cannot. That would surely be rare, but even so the “Required“ comment above isn't really true.
